File: docs/evidence/hair-2026-09-25/lock-wiring/summarize.py
from pathlib import Path
import gzip
import hashlib
import json
import math
import struct

root=Path('tmp/hair-sep25/lock-wiring')
out=root/'captures'
read=lambda p:json.loads(p.read_text())
sha=lambda data:hashlib.sha256(data).hexdigest()
report=read(out/'report.json')
results=report['results']
assert len(results)==21
for path,expected in report['sourceSHA256'].items():assert sha(Path(path).read_bytes())==expected,path
def bare(r):
    c=r['config']
    return f'{c["temporal"]}-{c["name"]}-p{c["phase"]}'
def name(r):return r['config']['wiring']+'-'+bare(r)
full={name(r):read(out/(name(r)+'.json')) for r in results}
prior_dir=Path('docs/evidence/hair-2026-09-25/resolve-finiteness')
prior={bare(r):r for r in map(json.loads,(prior_dir/'case-records.jsonl').read_text().splitlines())}
prior_rows={r['name']:r for r in read(prior_dir/'summary.json')['rows']}
rows=[]
for r in results:
    c,actual=r['config'],r['actual']
    wiring=c['wiring'];has_lock=wiring!='identity'
    assert actual['stats']['backend']=='webgpu' and actual['adapter']['vendor']=='apple'
    assert actual['stats']['temporalAA']=='taau' and actual['stats']['resolutionScale']==.66
    assert actual['frameId']==512 and actual['captureTime']==0
    assert actual['nativeAudit']['owner']=='TAAUNode' and not actual['nativeAudit']['sharpenNode']
    assert r['maskPixels']==12544 and r['bounds']=={'x0':72,'x1':184,'y0':72,'y1':184}
    targets=actual['nativeAudit']['targets']
    assert len(targets)==(4 if has_lock else 3)
    for t in targets:
        assert t['width']==t['height']==256 and t['gpuFormat']=='rgba16float' and t['typedArray']=='Uint16Array'
        assert t['attachments']==(1 if wiring=='identity' and t['name']=='resolve' else 2)
    assert len(r['routes'])==1
    route=r['routes'][0]
    assert route['arm']==wiring
    assert route['beforeSHA256']==sha((out/'original-served.js').read_bytes())
    assert route['afterSHA256']==sha((out/(wiring+'-served.js')).read_bytes())
    assert (route['beforeSHA256']==route['afterSHA256'])==(wiring=='identity')
    assert [f['frameId'] for f in r['frames']]==list(range(1,513))
    emitted_positive=history_positive=copy_different=0
    for f in r['frames']:
        a=f['nativeAudit']
        assert a['resolveHistoryBitDifferences']==0
        assert a['quantization']['finiteCompared']==262144 and a['quantization']['quantizationMismatches']==0
        for t in targets:
            for region,pixels in [('whole',65536),('roi',12544)]:
                v=a[t['name']][region]
                assert v['components']==pixels*4 and v['pixels']==v['redFinite']==pixels
                assert v['nan']==v['infinity']==v['nonfinitePixels']==0
        for i,offset in enumerate(f['offsets']):
            if c['mode']!='stochastic':continue
            rate=math.sqrt(2)-1 if c['sequence']=='distinct' and i==1 else .6180339887498949
            assert actual['materials'][i]['rate']==rate
            value=(f['frameId']+c['phase'])*rate
            assert offset==value-math.floor(value)
        if wiring=='copy-lock':assert a['lockHistoryBitDifferences']==0
        else:assert a['historyLock']['whole']['finiteMin']==a['historyLock']['whole']['finiteMax']==0
        if has_lock:
            emitted_positive+=a['resolveLock']['whole']['finiteMax']>0
            copy_different+=a['lockHistoryBitDifferences']>0
        history_positive+=a['historyLock']['whole']['finiteMax']>0
        if c['name'] in ['opaque-pair','empty']:
            expected=0 if c['name']=='opaque-pair' else 1
            assert f['min']==f['max']==expected
        if c['name']=='blend-half':assert abs(f['mean']-.25)<=1/255
    if has_lock and c['name']!='empty':assert emitted_positive>0
    if wiring=='attachment-only' and c['name']!='empty':assert copy_different>0 and history_positive==0
    if wiring=='copy-lock' and c['name']!='empty':assert history_positive>0 and copy_different==0
    pixel_sha=sha(struct.pack('<'+'d'*r['maskPixels'],*full[name(r)]['pixelMeans']))
    if wiring=='identity':
        assert r['frames']==prior[bare(r)]['frames']
        assert r['checkpoints']==prior[bare(r)]['checkpoints']
        assert pixel_sha==prior_rows[bare(r)]['pixelMeansSHA256']
    if wiring=='attachment-only':
        identity=full['identity-'+bare(r)]
        without_native=lambda x:[{k:v for k,v in f.items() if k!='nativeAudit'} for f in x['frames']]
        assert without_native(r)==without_native(identity)
        assert r['checkpoints']==identity['checkpoints']
        assert full[name(r)]['pixelMeans']==identity['pixelMeans']
    for s in r['checkpoints']:
        assert sha((out/f'{name(r)}-n{s["count"]}.png').read_bytes())==s['imageSHA256']
    tail=r['frames'][-64:]
    rows.append({'name':name(r),'nativeMetadata':actual['nativeAudit'],
      'emittedLockPositiveFrames':emitted_positive if has_lock else None,
      'historyLockPositiveFrames':history_positive,'lockCopyDifferentFrames':copy_different if has_lock else None,
      'nativeColorTailMean':sum(f['nativeAudit']['resolve']['roi']['redMean'] for f in tail)/64,
      'nativeHistoryLockTailMean':sum(f['nativeAudit']['historyLock']['roi']['redMean'] for f in tail)/64,
      'pixelMeansSHA256':pixel_sha,'checkpoints':r['checkpoints']})

by_name={name(r):r for r in results}
pilot=read(root/'pilot/report.json')
assert len(pilot['results'])==18 and pilot['sourceSHA256']==report['sourceSHA256']
for r in pilot['results']:
    later=by_name[name(r)]
    assert r['frames']==later['frames'][:24]
    assert r['checkpoints']==later['checkpoints'][:1]
    for route in r['routes']:
        assert route['beforeSHA256']==later['routes'][0]['beforeSHA256']
        assert route['afterSHA256']==later['routes'][0]['afterSHA256']
    assert sha((root/'pilot'/f'{name(r)}-n24.png').read_bytes())==r['checkpoints'][0]['imageSHA256']

summary={'baseCommit':report['baseCommit'],'sourceSHA256':report['sourceSHA256'],
 'finding':'Missing lock delivery confirmed in this fixture: a second resolve attachment emits lock but leaves history zero until explicitly copied. Copying transfers exact bits and changes color/noise; broader runtime qualification remains pending.',
 'cases':21,'framesPerCase':512,'pilotCasesExact':18,'priorIdentityCasesExact':7,'attachmentOnlyColorCasesExact':7,
 'rows':rows,'limits':['Static constant-alpha fixture, not real groom, fibre BSDF, motion or cost qualification.',
 'Three served-source arms; installed dependency and production source/assets unchanged.',
 'Native RGBA16F color/lock compared after copies; no assertion about every intermediate shader expression.',
 'A delivery repair does not qualify the rejected spatial sampler or make its remaining opacity bias acceptable.',
 'No threshold change or promotion; no full-suite/build/geometry/motion result.']}
(root/'summary.json').write_text(json.dumps(summary,indent=2)+'\n')
for label,records in [('case-records',results),('pilot-records',pilot['results'])]:
    data=''.join(json.dumps(r,separators=(',',':'))+'\n' for r in records).encode()
    (root/(label+'.jsonl.gz')).write_bytes(gzip.compress(data,mtime=0))
print('PASS: 21 actual Apple/Metal WebGPU cases x 512 frames; applied routes, attachments, formats, copies and native finiteness.')
print('PASS: seven identity cases reproduce prior frame traces, per-pixel means and 21 images exactly.')
print('PASS: seven attachment-only cases reproduce all identity color measurements, per-pixel means and images exactly.')
print('PASS: emitted nonzero locks remain absent from history without the copy; copied lock/history match bit-for-bit.')
print('PASS: eighteen pilot cases repeat all first-24 traces and images; endpoints/blend references remain valid.')
for row in rows:
    if 'half' in row['name'] and 'blend' not in row['name']:
        c=row['checkpoints'][-1]
        print(row['name'],'tail',c['tailMean'],'RMS',c['tailTemporalRMS'],'nativeLockTail',row['nativeHistoryLockTailMean'])
print(summary['finding'])

File: docs/evidence/hair-2026-09-25/lock-runtime/summarize.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import {createHash} from 'node:crypto';
import {gunzipSync} from 'node:zlib';
import {decodePng} from '../../../tools/critic/png.mjs';
import {patch} from './route.mjs';
const root='tmp/hair-sep25/lock-runtime';
const read=(p)=>fs.readFileSync(p),json=p=>JSON.parse(read(p));
const sha=b=>createHash('sha256').update(b).digest('hex');
const sourceHashes=json(root+'/source-hashes.json');
for(const [p,h] of Object.entries(sourceHashes))assert.equal(sha(read(p)),h,p);
const summary={sourceHashes,verdict:'Rejected for promotion: delivery works, T1 worsens; T1/T2/C4 remain red. No production change.',runs:{},baselineRepeats:[],pilotRepeats:[],maskComparisons:[],imageComparisons:[]};
const metric=(a,b,mask)=>{
  let pixels=0,changedPixels=0,squares=0,maxCodeValues=0;
  for(let p=0;p<a.width*a.height;p++){
    if(mask&&!mask[p])continue;
    pixels++;let changed=false;
    for(let c=0;c<3;c++){
      const d=Math.abs(Math.round(a.pixels[p*4+c]*255)-Math.round(b.pixels[p*4+c]*255));
      if(d)changed=true;squares+=d*d;maxCodeValues=Math.max(maxCodeValues,d);
    }
    if(changed)changedPixels++;
  }
  return {pixels,changedPixels,rmsCodeValues:pixels?Math.sqrt(squares/(3*pixels)):null,maxCodeValues};
};
for(const arm of ['identity','copy-lock'])for(const probe of ['tips','opacity']){
  const name=`${arm}-${probe}`,dir=root+'/'+name,pilot=root+'/pilot-'+name;
  const pipeline=json(dir+'/pipeline.json');
  assert.equal(pipeline.arm,arm);assert.equal(pipeline.routes.length,1);
  const before=read(dir+'/original-served.js'),after=read(dir+'/candidate-served.js');
  assert.equal(patch(before.toString(),arm),after.toString());
  assert.equal(pipeline.routes[0].beforeSHA256,sha(before));assert.equal(pipeline.routes[0].afterSHA256,sha(after));
  assert.equal(pipeline.snapshots.length,probe==='tips'?2:17);
  const native=[];
  pipeline.snapshots.forEach((s,i)=>{
    assert.equal(s.frameId,1+24*i);assert.equal(s.captureTime,0);
    assert.equal(s.stats.backend,'webgpu');assert.equal(s.stats.temporalAA,'taau');assert.equal(s.stats.resolutionScale,.66);
    assert.equal(s.stats.hairOIT,'stochastic');assert.equal(s.hairMaterialType,'HairNodeMaterial');
    assert.equal(s.phase,0);assert.equal(s.defect,null);assert.equal(s.hasInner,false);
    assert.equal(s.targets.length,arm==='identity'?3:4);
    assert.equal(s.copies.historyColor,0);
    if(arm==='copy-lock')assert.equal(s.copies.historyLock,0);
    for(const t of s.targets){
      assert.equal(t.width,900);assert.equal(t.height,1200);assert.equal(t.type,1016);assert.equal(t.gpuFormat,'rgba16float');
      assert.equal(t.strideBytes,7424);assert.equal(t.paddingBytes,224);
      assert.equal(t.values.components,4320000);assert.equal(t.values.nan+t.values.infinity,0);
      assert.equal(t.attachments,t.name==='color'?(arm==='identity'?1:2):2);
    }
    const lock=s.targets.find(t=>t.name==='historyLock').values;
    if(arm==='identity')assert.equal(lock.finiteMax,0);else assert.ok(lock.redMean>0);
    native.push({frame:s.frameId,lockMean:lock.redMean,lockMax:lock.finiteMax});
  });
  const measureNames=probe==='tips'?['measurements.json']:['portrait-measurements.json','rear34-measurements.json'];
  const measurements=Object.fromEntries(measureNames.map(n=>[n,json(dir+'/'+n)]));
  const images=fs.readdirSync(dir).filter(n=>n.endsWith('.png'));
  assert.equal(images.length,probe==='tips'?3:10);
  for(const n of [...images,...measureNames]){
    assert.deepEqual(read(dir+'/'+n),read(pilot+'/'+n),'Repeat '+name+'/'+n);
    summary.pilotRepeats.push({file:name+'/'+n,sha256:sha(read(dir+'/'+n))});
  }
  // Compare stable native records; timing, output paths and local Vite ports are not results.
  const prior=json(pilot+'/pipeline.json');
  for(let i=0;i<pipeline.snapshots.length;i++){
    for(const key of ['frameId','captureTime','owner','hairMaterialType','phase','defect','hasInner','targets','copies'])
      assert.deepEqual(pipeline.snapshots[i][key],prior.snapshots[i][key],name+' '+key);
  }
  summary.runs[name]={route:pipeline.routes[0],native,measurements,imageHashes:Object.fromEntries(images.map(n=>[n,sha(read(dir+'/'+n))]))};
  if(arm==='identity'){
    const baseline=probe==='tips'?'tmp/hair-sep25/pattern-phase/zero-tips':'tmp/hair-sep24/runtime-coverage/baseline-opacity';
    for(const n of [...images,...measureNames]){
      if(fs.existsSync(baseline+'/'+n))assert.deepEqual(read(dir+'/'+n),read(baseline+'/'+n),'Baseline '+n);
      else {
        const archived=json('docs/evidence/hair-2026-09-25/lock-runtime/summary.json').baselineRepeats;
        const record=archived.find(r=>r.file===name+'/'+n);
        assert.ok(record,'Missing recorded baseline '+n);
        assert.equal(sha(read(dir+'/'+n)),record.sha256,'Recorded baseline '+n);
      }
      summary.baselineRepeats.push({file:name+'/'+n,baseline:baseline+'/'+n,sha256:sha(read(dir+'/'+n))});
    }
  }
}
for(const [probe,views] of [['tips',['stochastic']],['opacity',['portrait','rear34']]])for(const view of views){
  const dirs=['identity','copy-lock'].map(arm=>root+'/'+arm+'-'+probe);
  const masks=dirs.map(d=>JSON.parse(gunzipSync(read(d+'/'+view+'-masks.json.gz'))));
  assert.deepEqual(masks[0].raster,masks[1].raster,view+' geometry raster');
  assert.deepEqual(masks[0].regions,masks[1].regions,view+' final mask membership');
  summary.maskComparisons.push({probe,view,rasterSHA256:sha(JSON.stringify(masks[0].raster)),regionSHA256:sha(JSON.stringify(masks[0].regions)),exactMembership:true});
  const files=probe==='tips'?[view+'.png']:[view+'-hair.png',view+'-bald.png'];
  for(const file of files){
    const [a,b]=dirs.map(d=>decodePng(read(d+'/'+file)));
    assert.equal(a.width,900);assert.equal(a.height,1200);assert.equal(b.width,900);assert.equal(b.height,1200);
    const regions=Object.fromEntries(Object.entries(masks[0].regions).map(([k,v])=>[k,Buffer.from(v,'base64')]));
    const differences={whole:metric(a,b)};
    if(probe==='tips'){
      const boundary=regions.stable.map(x=>1-x);
      for(const [k,v] of Object.entries({...regions,boundary}))differences[k]=metric(a,b,v);
    }
    summary.imageComparisons.push({probe,file,regions:differences});
  }
}
fs.writeFileSync(root+'/summary.json',JSON.stringify(summary,null,2)+'\n');
console.log('PASS: source hashes, exact routes, native finite/copy delivery, 24-step clocks, all pilot results, shipping baselines and geometry/mask membership verified.');
console.log(JSON.stringify({repeatedFiles:summary.pilotRepeats.length,baselineFiles:summary.baselineRepeats.length,maskComparisons:summary.maskComparisons,imageComparisons:summary.imageComparisons},null,2));
